oomparser: update to use kmsg based parser

euank opened the pull request:
My comment for PR https://github.com/kubernetes/node-problem-detector/pull/41#issue-189041129 applies here too.

This would fix https://github.com/kubernetes/kubernetes/issues/34965, probably fix #1484, and work-around/fix #645.

Things that should be considered before merging:

1) anywhere kubelet or cadvisor runs in a container now needs access to `/dev/kmsg`. This should be documented as a breaking change and various examples / docs might need changing.
2) this reduces a little how far back we can see (namely after the system reboots, we don't get things before that). This implies that a "kubelet crash -> container OOM -> node reboot" chain of events can lose that container OOM.
3) this works everywhere where the kernel is modern enough to run docker! Fewer dependencies!


Testing note: I removed the log-file formatted files. The parsing of log-files is now the responsibility of the kmsgparser package which lives elsewhere, so we can trust it to do the right thing. We also get to assume the timestamp etc.
I have not done manual testing of this change, but I have done decent manual testing of the kmsgparser.

timstclair commented on utils/oomparser/oomparser.go:
```diff
@@ -101,132 +90,46 @@ func checkIfStartOfOomMessages(line string) bool {
-	}()
+// StreamOoms writes to a provided a stream of OomInstance objects representing
+// OOM events that are found in the logs.
+// It will blocks and should be called from a goroutine.
```
nit: s/blocks/block/

timstclair commented on utils/oomparser/oomparser.go:
```diff
@@ -101,132 +90,46 @@ func checkIfStartOfOomMessages(line string) bool {
-		ioreader: bufio.NewReader(tail),
-	}, nil
+	// Should not happen
+	glog.Warningf("exiting analyzeLines. OOM events will not be reported.")
```
nit: If this really should not happen it should be error level

euank replied:
Before it was info and I thought it should be error, so I picked the middle as a compromise :innocent:

timstclair commented on utils/oomparser/oomparser_test.go:
```diff
@@ -126,42 +118,203 @@ func TestCheckIfStartOfMessages(t *testing.T) {
+					"[ 1532]  1020  1532   410347   398810     788        0             0 badsysprogram",
+					"Out of memory: Kill process 1532 (badsysprogram) score 919 or sacrifice child",
+					"Killed process 1532 (badsysprogram) total-vm:1641388kB, anon-rss:1595164kB, file-rss:76kB",
+				},
```
Add a test case with multiple OOM events (note: I think writeAll needs to run in a goroutine to prevent deadlock)

euank replied:
Done

timstclair commented on utils/oomparser/oomparser.go:
```diff
@@ -101,132 +90,46 @@ func checkIfStartOfOomMessages(line string) bool {
-	parser, err = tryLogFile()
-	if err == nil {
-		return parser, nil
+	parser, err := kmsgparser.NewParser()
```
nit: I like the logger pattern, consider setting the logger to use glog for consistency?

timstclair commented:
Nice! This LGTM, but I'd like to get input from a few more people who understand the OOM pathways better than I do.

timstclair commented:
@vishh - Could you take a look at this?

euank commented on utils/oomparser/oomparser.go:
```diff
@@ -101,132 +90,46 @@ func checkIfStartOfOomMessages(line string) bool {
+// OOM events that are found in the logs.
+// It will blocks and should be called from a goroutine.
+func (self *OomParser) StreamOoms(outStream chan<- *OomInstance) {
+	kmsgEntries := self.parser.Parse()
```
Question: Should we call `self.parser.SeekEnd()` here or not?

I think `SeekEnd` more closely matches the `journalctl .. -f` behavior of before (start at the end of kmsg), but I think that looking a little into the past makes the kubelet behave more correctly after a restart.

Will OOM events that happen while the kubelet is down, but are detected on kubelet startup actually do the right thing?
Are they idempotent so that power-cycling the kubelet having bunches of duplicate events is okay?

timstclair replied:
I think reporting recent OOM events is desireable, and it's also consistent with the non-journald behavior.

euank commented:
@timstclair addressed your comments and fixed a bug where OOM events with a process name containing a `-` in them would break the parser (wouldn't match the end-line for that oom). Lucky me that my test program has a dash :smile:

euank commented:
Rebased and squashed the commit addressing comments from back whenever.

goettl79 commented:
This fix  would also make the packaging issue of cadvisor / journalctl inside a docker container obsolete.  https://github.com/google/cadvisor/issues/1313

euank commented:
Bump on this, did a rote rebase and afaict there aren't open concerns (other than waiting on whether @vishh feels like reviewing)

euank commented:
@k8s-bot test this

dchen1107 commented:
cc/ @timstclair @dashpole @vishh

euank commented:
Rebased for probably the 10th or so time.

Manually tested via the following:

```sh
$ ./cadvisor -v 3 -logtostderr
...
# cgroup OOM
$ docker run -m 15M euank/gunpowder-memhog:latest 30M
# cadvisor logs:
I0830 20:24:01.875221    3264 manager.go:1142] Created an OOM event in container "/docker/6da85dc9675db12149e970c6ef6835240bab712e5499963d13e88a446248e82b" at 2017-08-30 20:24:01.583571597 +0000 UTC

# system OOM
$ docker run -d euank/gunpowder-memhog 500M
$ sleep 5
$ echo f | sudo tee /proc/sysrq-trigger
# cadvisor logs
I0830 20:59:27.831699    1611 manager.go:1142] Created an OOM event in container "/" at 2017-08-30 20:59:27.717686391 +0000 UTC
```

@timstclair if this is going to land, can we land it?
Having to come by and rebase it once a month is ridiculous.

vishh commented:
@euank I can help review this. is this ready for a review?

euank commented:
@vishh yup, it's been ready for review for 9 months now.

euank commented:
🎉
